### PR TITLE
Synchronize ufal and dataquest main branches

### PR DESCRIPTION
**Pull request by kosarko**
This contains the following fixes:

Issue 1165: Handle PaginantionException in Clarin licences search requests (ufal/clarin-dspace#1184)
Issue 1055: fixed security issue for downloading file with non anonymous license (ufal/clarin-dspace#1188)
Issue 1186: catch DSpaceBadRequestException rather than BadRequestException (ufal/clarin-dspace#1187)
Issue ufal/clarin-dspace#1190 - fixed Server error  (ufal/clarin-dspace#1193)
checkstyle
Shortref v7 (ufal/clarin-dspace#1194)
remove duplicate dependency element with identical content (warnings in build)

## Summary by CodeRabbit

- **New Features**
  - Added methods for counting handles and finding handles by token with enhanced validation.
  - Introduced a random string generator component to improve handle creation reliability.
  - Enhanced Clarin License confirmation handling using a new enum for clearer state representation.

- **Bug Fixes**
  - Improved error responses and logging for handle operations and license pagination.
  - Fixed regex pattern matching in handle validation and improved null safety in bitstream processing.

- **Refactor**
  - Replaced raw integer license confirmation values with descriptive enum constants across code and tests.
  - Standardized exception handling by replacing generic bad request exceptions with a custom application-specific exception.
  - Improved error logging strategy to differentiate client and server errors with configurable logging levels.
  - Updated handle URL generation to include fresh tokens and improved authorization handling during updates.

- **Tests**
  - Expanded integration tests for handle management and Clarin License workflows, including token validation and blacklist enforcement.
  - Added tests for pagination edge cases and unauthorized access scenarios based on license confirmation states.
  - Enhanced test setup and cleanup to maintain environment consistency.

- **Chores**
  - Removed duplicate dependencies and unused imports.
  - Updated configuration to enforce stricter URL and host validation rules for handle shortener functionality.

### CODE REVIEW

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/RandomStringGeneratorImpl.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.app.rest.repository;
+
+import org.apache.commons.lang.RandomStringUtils;
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/RandomStringGeneratorImpl.java]: Following the code rabit review:
```
The code is using the older, deprecated version of Apache Commons Lang. This version is no longer maintained and may have security vulnerabilities.

-import org.apache.commons.lang.RandomStringUtils;
+import org.apache.commons.lang3.RandomStringUtils;

Consider updating the dependency in your build configuration to use the newer version.
```

**Review by milanmajchrak — CHANGES_REQUESTED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ExternalHandleRestRepository.java [hunk 2]
```diff
@@ -257,7 +279,7 @@ private boolean matchesAnyOf(String tested, String configPropertyWithPatterns) {
             return false;
         }
 
-        final String pattern = String.join(",", patterns);
+        final String pattern = String.join(";", patterns);
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ExternalHandleRestRepository.java:282]: This is probably not correct:
shortener.post.url.blacklist.regexps = .*://.{1,15}

matchesAnyOf(host, "shortener.post.host.blacklist.regexps")

getArrayProperty splits by `,` you join it with `;` and split again with `;`

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java [hunk 3]
```diff
@@ -87,15 +86,15 @@ public Page<MetadataBitstreamWrapperRest> findByHandle(@Parameter(value = "handl
...
         List<Bundle> bundles = findEnabledBundles(fileGrpTypes, item);
         for (Bundle bundle : bundles) {
-            List<Bitstream> bitstreams = bundle.getBitstreams();
+            List<Bitstream> bitstreams = new ArrayList<>(bundle.getBitstreams());
             String use = bundle.getName();
             if (StringUtils.equals("THUMBNAIL", use)) {
                 Thumbnail thumbnail = itemService.getThumbnail(context, item, false);
                 if (Objects.nonNull(thumbnail)) {
-                    bitstreams = new ArrayList<>();
+                    bitstreams.clear();
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java:101]: what exactly should this do?

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java [hunk 4]
```diff
@@ -109,32 +108,30 @@ public Page<MetadataBitstreamWrapperRest> findByHandle(@Parameter(value = "handl
                         List<PreviewContent> prContents = previewContentService.hasPreview(context, bitstream);
                         // Generate new content if we didn't find any
                         if (prContents.isEmpty()) {
-                            fileInfos = previewContentService.getFilePreviewContent(context, bitstream);
+                            fileInfos.addAll(previewContentService.getFilePreviewContent(context, bitstream));
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java]: why this change?

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java (hunk 3 above)
Comment (kuchtiak-ufal) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java:101]: This is just refactoring, due to some warning (in IntelliG) about reassigning local variables.

Instead of
```
List<Bitstream> bitstreams = bundle.getBitstreams();
if (condition) {
   bitstream = new ArrayList();
}
```

There is now:
```
 List<Bitstream> bitstreams = new ArrayList<>(bundle.getBitstreams());
 if (condition) {
    bitstream.clear();
 }
```

There is also no danger now that if something happens with **bitstreams** list - that would influence **bundle.bitstreams** list;

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java (hunk 4 above)
Comment (kuchtiak-ufal) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java]: Similar as above -> avoid reassigning local variables.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ExternalHandleRestRepository.java (hunk 2 above)
Comment (kosarko) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ExternalHandleRestRepository.java:282]: @vidiecan 
see the hostBlacklist test, I'm hoping that one is testing the config correctly.

getArrayProperty works in two "modes"
```
a=1
a=2
a=3
```
and 
```
a=1,2,3
```
and these should produce the same result `["1", "2", "3"]`.

In v5 we've used `;` for a similar purpose - to define an array of patterns. I'm assuming we are keeping that in v7 for compatibility. 

But what happens if you mix the two approaches?

```
a=1
a=2;3
```
getArrayProperty will load that as `["1", "2;3"]`? If you join that with `,` and then split with `;` you get `["1,2", "3"]`, but I'd say the correct result is `["1", "2", "3"]` and you get that with this changes, at least I hope you do.

Comment (kosarko) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ExternalHandleRestRepository.java:282]: >shortener.post.url.blacklist.regexps = .*://.{1,15}

I see, it's broken when there's a comma in the regexep

Comment (kosarko) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ExternalHandleRestRepository.java:282]: @vidiecan Seems that the config value is always split on commas (no difference between `getArrayProperty` and `getProperty`), so the reasonable thing to do here is escape the commas where you don't want the split to happen. I've added an example of that into clarin-dspace.cfg.

Also this is probably a lindat only feature and currently we are (in production) blacklisting everything but 3 domains.

File: dspace-server-webapp/pom.xml [hunk 5]
```diff
@@ -290,12 +290,6 @@
             <version>${spring-boot.version}</version>
         </dependency>
 
-        <dependency>
```
Comment (vidiecan) [inline dspace-server-webapp/pom.xml:293]: are we sure about this?

what about

https://github.com/dataquest-dev/DSpace/blob/6daa0fdd88e4d377b16c01ec70430c68b89bb06e/dspace-server-webapp/src/main/java/org/dspace/app/rest/configuration/ActuatorConfiguration.java#L23

Comment (kosarko) [reply dspace-server-webapp/pom.xml:293]: @vidiecan yes, see the lines above this: L287-L291
this is originally from https://github.com/ufal/clarin-dspace/pull/1138/files

**Review by vidiecan — APPROVED** [review]

**Review by milanmajchrak — APPROVED** [review]
